src/VisualStudio/Core/Def/Snippets/RoslynLSPSnippetParser.cs

```diff
@@ -0,0 +1,250 @@
+{
+    /*
+     * LSP's full snippet syntax defined in the following EBNF grammar.
+     * -------------------------------------------------------------------------------------------------------------------------
+     * any         ::= tabstop | placeholder | choice | variable | text
+     * tabstop     ::= '$' int | '${' int '}'
+     * placeholder ::= '${' int ':' any '}'
+     * choice      ::= '${' int '|' text (',' text)* '|}'
+     * variable    ::= '$' var | '${' var }'
+     *                 | '${' var ':' any '}'
+     *                 | '${' var '/' regex '/' (format | text)+ '/' options '}'
+     * format      ::= '$' int | '${' int '}'
+     *                 | '${' int ':' '/upcase' | '/downcase' | '/capitalize' '}'
+     *                 | '${' int ':+' if '}'
+     *                 | '${' int ':?' if ':' else '}'
+     *                 | '${' int ':-' else '}' | '${' int ':' else '}'
+     * regex       ::= Regular Expression value (ctor-string)
+     * options     ::= Regular Expression option (ctor-options)
+     * var         ::= [_a-zA-Z] [_a-zA-Z0-9]*
+     * int         ::= [0-9]+
+     * text        ::= .*
+     * -------------------------------------------------------------------------------------------------------------------------
+     * We don't currently support all of the above, the current parsers only supports text, tabstop and non-nested placeholders:
+     * -------------------------------------------------------------------------------------------------------------------------
+     * any         ::= tabstop | placeholder | text
+     * tabstop     ::= '$' int | '${' int '}'
+     * placeholder ::= '${' int ':' text '}'
+     * int         ::= [0-9]+
+     * text        ::= .*
+     */
+
+    private const string TabStopIndexGroupName = "tabStopIndex";
+    private const string PlaceholderNameGroupName = "placeholderName";
+    private const string TextGroupName = "text";
+    private const string TabStopGroupName = "tabstop";
+    private const string PlaceholderGroupName = "placeholder";
+    private const string TextRegexPattern = "[^$]+";
+
+    // tabstop     ::= '$' int | '${' int '}'
+    private const string TabStopRegexPattern = $@"\$(?<{TabStopIndexGroupName}>[0-9]+)|\$\{{(?<{TabStopIndexGroupName}>[0-9]+)\}}";
+
+    // placeholder ::= '${' int ':' text '}'
+    private const string PlaceholderRegexPattern = $@"\$\{{(?<{TabStopIndexGroupName}>[0-9]+):(?<{PlaceholderNameGroupName}>.+?)\}}";
+    private const string SnippetRegexPattern = $@"(?<{TabStopGroupName}>{TabStopRegexPattern})|(?<{PlaceholderGroupName}>{PlaceholderRegexPattern}|(?<{TextGroupName}>{TextRegexPattern}))";
+    private static readonly Regex s_snippetRegex = new(SnippetRegexPattern, RegexOptions.Compiled | RegexOptions.Singleline, TimeSpan.FromSeconds(1));
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public RoslynLSPSnippetParser()
```

Comment:
Why is this MEF versus just a static class?

Reply:
It was simply preserved from the original implementation.